From 8de0d0c12fe7e512a8183d2b74c1f967ee1b2674 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:20:56 -0400
Subject: [net] netfilter: nf_tables: wait for call_rcu completion on module removal

Message-id: <20141021162056.22888.40502.stgit@localhost.localdomain>
Patchwork-id: 98009
O-Subject: [RHEL7 net PATCH 112/124] netfilter: nf_tables: wait for call_rcu completion on module removal
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 1b1bc49c0fc0501bf0d1366a2a5e5c1f8dcf9cb1

 commit 1b1bc49c0fc0501bf0d1366a2a5e5c1f8dcf9cb1
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Wed Oct 1 13:53:20 2014 +0200

    netfilter: nf_tables: wait for call_rcu completion on module removal

    Make sure the objects have been released before the nf_tables modules
    is removed.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index c0831f6..7053c70 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -4163,6 +4163,7 @@ static void __exit nf_tables_module_exit(void)
 {
  unregister_pernet_subsys(&nf_tables_net_ops);
  nfnetlink_subsys_unregister(&nf_tables_subsys);
+ rcu_barrier();
  nf_tables_core_module_exit();
  kfree(info);
 }
-- 
1.7.1